### PR TITLE
Log useful information on 500's; include error message in error (!!)

### PR DESCRIPTION
**Pull request by lavalamp**
Diagnose #31728

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31756)

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
Lgtm - this is a big help

**Comment by lavalamp** [comment]
ok to test

@ixdy why does the pull test ask this on every PR, even mine?

**Comment by ixdy** [comment]
I blame @fejta 

(https://github.com/kubernetes/test-infra/pull/459)

**Comment by lavalamp** [comment]
test should work now.

@derekwaynecarr do you want to make that lgtm official? ;)

**Comment by smarterclayton** [comment]
Tiny security risk here (leaking potentially internal messages) but I think this is strictly more useful than the alternative.

**Comment by lavalamp** [comment]
Yeah. But right now we don't even *log* the internal error messages anywhere.

**Comment by smarterclayton** [comment]
I thought we did via the error handler - we should be reporting it to
util.HandleError whenever we get a non-APIStatus style error or an internal
server error.

**Comment by smarterclayton** [comment]
pkg/apiserver/errors.go logs errors returned by REST interfaces that are
not in the right form - but I think we need to log any InternalServerError
(there is no case we should return that without logging as well).  We could
log it in that method or at higher (i slightly prefer here because it is
tied together).

**Comment by lavalamp** [comment]
Empirically, we don't put anything useful in the log file other than the stack trace, and that isn't useful enough.

**Comment by smarterclayton** [comment]
I think 90% of the issues we've found at runtime have been through HandleError.  We flag 2-3 a day from our production clusters which are instrumented with sentry.  Agree it could be better in the default log output, but I would definitely like to see messages show up and correlate to time.

File: pkg/api/errors/errors.go [hunk 1]
```diff
@@ -325,7 +325,7 @@ func NewGenericServerResponse(code int, verb string, qualifiedResource unversion
 	default:
 		if code >= 500 {
 			reason = unversioned.StatusReasonInternalError
-			message = "an error on the server has prevented the request from succeeding"
+			message = fmt.Sprintf("an error on the server (%q) has prevented the request from succeeding", serverMessage)
```
Comment (smarterclayton) [inline pkg/api/errors/errors.go]: I didn't catch this but server message could be many KB long

**Pull request merged**